File: sdk-javascript/src/registry/RegistryClient.ts
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

import { X509Certificate, createPublicKey } from 'node:crypto';

import type { KeyLike } from 'jose';
import { LRUCache } from 'lru-cache';

import { JWE_ALG } from '../crypto/JweAlgorithms.js';
import {
  ParticipantNotFoundError,
  RegistryUnavailableError,
  TransportError,
} from '../exceptions/HfcxError.js';
import type { ParticipantCert, RecipientCertResolver } from './ParticipantCert.js';

import type { FetchFn } from '../auth/KeycloakTokenClient.js';

export interface RegistryClientOptions {
  /**
   * Sunbird-RC participant-registry root, e.g.
   * `https://registry.hcx-egypt.gov.eg`. Lookup paths are appended at
   * request time.
   */
  baseUrl: string | URL;
  /** Time before `notAfter` to evict an entry. Default 1 h. */
  preExpiryBufferMs?: number;
  /** LRU bound. Default 10 000. */
  maxEntries?: number;
  /** Per-request timeout. Default 10 s. */
  requestTimeoutMs?: number;
  /** Test seam: inject a custom fetch (defaults to `globalThis.fetch`). */
  fetchFn?: FetchFn;
  /** Test seam: wall clock. */
  clock?: () => number;
}

interface CacheEntry {
  readonly cert: ParticipantCert;
  readonly expiresAtMs: number;
}

/**
 * LRU-cached lookup of HFCX participant encryption certs against the
 * platform's Sunbird-RC participant registry. Cross-SDK invariants
 * identical to Java's `RegistryClient`, Python's `AsyncRegistryClient`,
 * and .NET's `RegistryClient`.
 */
export class RegistryClient implements RecipientCertResolver {
  /** Default time before `notAfter` to evict — 1 hour. */
  static readonly DEFAULT_PRE_EXPIRY_BUFFER_MS = 60 * 60 * 1000;

  /** Default LRU cache capacity. */
  static readonly DEFAULT_MAX_ENTRIES = 10_000;

  private readonly baseUrl: URL;
  private readonly preExpiryBufferMs: number;
  private readonly requestTimeoutMs: number;
  private readonly fetchFn: FetchFn;
  private readonly clock: () => number;
  private readonly cache: LRUCache<string, CacheEntry>;

  constructor(options: RegistryClientOptions) {
    if (!options.baseUrl) throw new TypeError('baseUrl is required');
    this.baseUrl = options.baseUrl instanceof URL ? options.baseUrl : new URL(options.baseUrl);
    this.preExpiryBufferMs =
      options.preExpiryBufferMs ?? RegistryClient.DEFAULT_PRE_EXPIRY_BUFFER_MS;
    this.requestTimeoutMs = options.requestTimeoutMs ?? 10_000;
    this.fetchFn = options.fetchFn ?? globalThis.fetch;
    this.clock = options.clock ?? Date.now;
    this.cache = new LRUCache<string, CacheEntry>({
      max: options.maxEntries ?? RegistryClient.DEFAULT_MAX_ENTRIES,
    });
  }

  async getRecipientCert(participantCode: string): Promise<ParticipantCert> {
    if (!participantCode) throw new TypeError('participantCode is required');

    const cached = this.cache.get(participantCode);
    if (cached && cached.expiresAtMs > this.clock()) {
      return cached.cert;
    }

    const fetched = await this.fetchOnce(participantCode);
    const ttl = fetched.notAfter.getTime() - this.preExpiryBufferMs - this.clock();
    if (ttl > 0) {
      this.cache.set(
        participantCode,
        { cert: fetched, expiresAtMs: this.clock() + ttl },
        {
          ttl,
        },
      );
    }
    return fetched;
  }

  /** Drop the cached entry for one participant. */
  invalidate(participantCode: string): void {
    this.cache.delete(participantCode);
  }

  /** Drop every cached entry. Idempotent. */
  invalidateAll(): void {
    this.cache.clear();
  }

  private async fetchOnce(participantCode: string): Promise<ParticipantCert> {
    const url = new URL('/api/v1/Participant/search', this.baseUrl);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.requestTimeoutMs);

    let response: Response;
    try {
      response = await this.fetchFn(url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
        body: JSON.stringify({
          filters: { participant_code: { eq: participantCode } },
        }),
        signal: controller.signal,
      });
    } catch (err) {
      throw new RegistryUnavailableError(
        `Registry unreachable for ${participantCode}: ${(err as Error).message ?? String(err)}`,
      );
    } finally {
      clearTimeout(timer);
    }

    if (response.status === 404) {
      throw new ParticipantNotFoundError(`Registry has no entry for ${participantCode}`);
    }
    if (response.status >= 500) {
      throw new RegistryUnavailableError(
        `Registry returned HTTP ${response.status} for ${participantCode}`,
      );
    }
    if (!response.ok) {
      throw new TransportError(`Registry returned HTTP ${response.status} for ${participantCode}`);
    }

    const body = await response.text();
    return parseRegistryEntry(participantCode, body);
  }
}

function parseRegistryEntry(participantCode: string, body: string): ParticipantCert {
  let root: unknown;
  try {
    root = JSON.parse(body);
  } catch (err) {
    throw new TransportError(
      `Registry response for ${participantCode} was not valid JSON: ${(err as Error).message}`,
    );
  }

  let entries: unknown;
  if (Array.isArray(root)) {
    entries = root;
  } else if (
    root !== null &&
    typeof root === 'object' &&
    Array.isArray((root as { entity?: unknown }).entity)
  ) {
    entries = (root as { entity: unknown[] }).entity;
  } else {
    throw new ParticipantNotFoundError(`Registry response had no entry for ${participantCode}`);
  }

  const list = entries as unknown[];
  if (list.length === 0) {
    throw new ParticipantNotFoundError(`Registry response had no entry for ${participantCode}`);
  }

  const first = list[0];
  if (
    !first ||
    typeof first !== 'object' ||
    typeof (first as { encryption_cert?: unknown }).encryption_cert !== 'string'
  ) {
    throw new TransportError(`Registry entry for ${participantCode} has no encryption_cert`);
  }

  const pem = (first as { encryption_cert: string }).encryption_cert;
  return parseCert(participantCode, pem);
}

function parseCert(participantCode: string, pem: string): ParticipantCert {
  let cert: X509Certificate;
  try {
    cert = new X509Certificate(pem);
  } catch (err) {
    throw new TransportError(
      `Failed to parse encryption_cert PEM for ${participantCode}: ${(err as Error).message}`,
    );
  }

  const publicKey = cert.publicKey;
  if (publicKey.asymmetricKeyType !== 'rsa') {
    throw new TransportError(
      `encryption_cert for ${participantCode} is not RSA (got ${publicKey.asymmetricKeyType ?? 'unknown'})`,
    );
  }

  // Re-import via SPKI bytes to materialise a `jose`-friendly KeyObject
  // tagged with the pinned algorithm.
  const spki = createPublicKey({
    key: publicKey.export({ type: 'spki', format: 'der' }),
    format: 'der',
    type: 'spki',
  });
  // jose's KeyLike accepts node KeyObject directly when it has the right
  // alg metadata; passing it through createPublicKey ensures a fresh,
  // detached copy.
  const keyLike = spki as unknown as KeyLike;
  // Touch JWE_ALG so unused-import linters don't strip the symbol — it
  // documents which algorithm this key is intended for.
  void JWE_ALG;

  return {
    participantCode,
    publicKey: keyLike,
    notAfter: new Date(cert.validTo),
  };
}

File: sdk-dotnet/src/HealthFlow.Hfcx.Sdk/Registry/RegistryClient.cs
// Copyright (c) HealthFlow Medical HCX. Licensed under Apache 2.0.

using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Security.Cryptography.X509Certificates;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using HealthFlow.Hfcx.Sdk.Exceptions;
using Microsoft.Extensions.Caching.Memory;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace HealthFlow.Hfcx.Sdk.Registry;

/// <summary>
/// Caffeine-cached lookup of HFCX participant encryption certs against
/// the platform's Sunbird-RC participant registry.
/// </summary>
/// <remarks>
/// <para>Sister to the Java SDK's <c>RegistryClient</c> and the Python
/// SDK's <c>AsyncRegistryClient</c>. Cross-SDK invariants:</para>
/// <list type="bullet">
///   <item>Per-entry TTL = cert <c>notAfter - PreExpiryBuffer</c>
///     (default 1h).</item>
///   <item>Bounded by an LRU cache; default 10 000 entries.</item>
///   <item>404 → <see cref="ParticipantNotFoundException"/>.</item>
///   <item>5xx / network failures → <see cref="RegistryUnavailableException"/>.</item>
///   <item>Malformed JSON / PEM / non-RSA cert →
///     <see cref="TransportException"/>.</item>
///   <item>Hit / miss / eviction stats logged at INFO at most every 60s.</item>
/// </list>
/// </remarks>
public sealed class RegistryClient : IRecipientCertResolver, IDisposable
{
    /// <summary>Default time before <c>notAfter</c> to evict — 1 hour.</summary>
    public static readonly TimeSpan DefaultPreExpiryBuffer = TimeSpan.FromHours(1);

    /// <summary>Default LRU cache capacity.</summary>
    public const int DefaultMaxEntries = 10_000;

    private readonly Uri _baseUrl;
    private readonly HttpClient _httpClient;
    private readonly bool _ownsHttpClient;
    private readonly TimeSpan _preExpiryBuffer;
    private readonly MemoryCache _cache;
    private readonly Func<DateTimeOffset> _clock;
    private readonly ILogger _log;
    private readonly object _statsLock = new();
    private long _hits;
    private long _misses;
    private long _evictions;
    private DateTimeOffset _lastStatsLog;

    /// <summary>
    /// Construct a <see cref="RegistryClient"/>.
    /// </summary>
    /// <param name="baseUrl">
    /// Sunbird-RC participant-registry root, e.g.
    /// <c>https://registry.hcx-egypt.gov.eg</c>. Lookup paths are
    /// appended at request time.
    /// </param>
    /// <param name="httpClient">
    /// Optional injected <see cref="HttpClient"/>. When omitted the
    /// client owns and disposes a fresh instance.
    /// </param>
    /// <param name="preExpiryBuffer">
    /// Time before <c>notAfter</c> to evict an entry. Default 1h.
    /// </param>
    /// <param name="maxEntries">LRU bound. Default 10 000.</param>
    /// <param name="logger">Optional logger.</param>
    /// <param name="clock">Test seam for the wall clock.</param>
    public RegistryClient(
        Uri baseUrl,
        HttpClient? httpClient = null,
        TimeSpan? preExpiryBuffer = null,
        int? maxEntries = null,
        ILogger<RegistryClient>? logger = null,
        Func<DateTimeOffset>? clock = null)
    {
        ArgumentNullException.ThrowIfNull(baseUrl);

        _baseUrl = baseUrl;
        _preExpiryBuffer = preExpiryBuffer ?? DefaultPreExpiryBuffer;

        if (httpClient is null)
        {
            _httpClient = new HttpClient();
            _ownsHttpClient = true;
        }
        else
        {
            _httpClient = httpClient;
            _ownsHttpClient = false;
        }

        var cap = maxEntries ?? DefaultMaxEntries;
        _cache = new MemoryCache(new MemoryCacheOptions
        {
            SizeLimit = cap,
            CompactionPercentage = 0.1,
        });
        _log = logger ?? NullLogger<RegistryClient>.Instance;
        _clock = clock ?? (() => DateTimeOffset.UtcNow);
        _lastStatsLog = _clock();
    }

    /// <inheritdoc />
    public async Task<ParticipantCert> GetRecipientCertAsync(
        string participantCode,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(participantCode);

        if (_cache.TryGetValue<ParticipantCert>(participantCode, out var cached) && cached is not null)
        {
            Interlocked.Increment(ref _hits);
            MaybeLogStats();
            return cached;
        }

        Interlocked.Increment(ref _misses);

        var fetched = await FetchAsync(participantCode, cancellationToken).ConfigureAwait(false);
        var ttl = TtlFor(fetched);
        if (ttl <= TimeSpan.Zero)
        {
            // Cert is already inside its pre-expiry buffer — return it but
            // do not cache; the next call will fetch a fresh one.
            return fetched;
        }

        _cache.Set(
            participantCode,
            fetched,
            new MemoryCacheEntryOptions
            {
                Size = 1,
                AbsoluteExpirationRelativeToNow = ttl,
                PostEvictionCallbacks =
                {
                    new PostEvictionCallbackRegistration
                    {
                        EvictionCallback = (_, _, reason, _) =>
                        {
                            if (reason is EvictionReason.Capacity or EvictionReason.Expired)
                            {
                                Interlocked.Increment(ref _evictions);
                            }
                        },
                    },
                },
            });
        MaybeLogStats();
        return fetched;
    }

    /// <summary>Drop the cached entry for one participant.</summary>
    public void Invalidate(string participantCode)
    {
        ArgumentException.ThrowIfNullOrEmpty(participantCode);
        _cache.Remove(participantCode);
    }

    /// <summary>Drop every cached entry. Idempotent.</summary>
    public void InvalidateAll()
    {
        _cache.Compact(1.0);
    }

    private TimeSpan TtlFor(ParticipantCert cert)
        => cert.NotAfter - _preExpiryBuffer - _clock();

    private async Task<ParticipantCert> FetchAsync(
        string participantCode,
        CancellationToken cancellationToken)
    {
        var path = $"/api/v1/Participant/search";
        var requestUri = new Uri(_baseUrl, path);

        HttpResponseMessage? response = null;
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, requestUri);
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
            request.Content = new StringContent(
                $"{{\"filters\":{{\"participant_code\":{{\"eq\":\"{participantCode}\"}}}}}}",
                System.Text.Encoding.UTF8,
                "application/json");

            response = await _httpClient
                .SendAsync(request, HttpCompletionOption.ResponseContentRead, cancellationToken)
                .ConfigureAwait(false);

            if (response.StatusCode == HttpStatusCode.NotFound)
            {
                throw new ParticipantNotFoundException(
                    $"Registry has no entry for {participantCode}");
            }

            if ((int)response.StatusCode >= 500)
            {
                throw new RegistryUnavailableException(
                    $"Registry returned HTTP {(int)response.StatusCode} for {participantCode}");
            }

            if (!response.IsSuccessStatusCode)
            {
                throw new TransportException(
                    $"Registry returned HTTP {(int)response.StatusCode} for {participantCode}");
            }

            var body = await response.Content
                .ReadAsStringAsync(cancellationToken)
                .ConfigureAwait(false);
            return ParseRegistryEntry(participantCode, body);
        }
        catch (HttpRequestException ex)
        {
            throw new RegistryUnavailableException(
                $"Registry unreachable for {participantCode}: {ex.Message}");
        }
        catch (TaskCanceledException ex) when (!cancellationToken.IsCancellationRequested)
        {
            throw new RegistryUnavailableException(
                $"Registry request for {participantCode} timed out: {ex.Message}");
        }
        finally
        {
            response?.Dispose();
        }
    }

    private static ParticipantCert ParseRegistryEntry(string participantCode, string body)
    {
        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(body);
        }
        catch (JsonException ex)
        {
            throw new TransportException(
                $"Registry response for {participantCode} was not valid JSON: {ex.Message}");
        }

        using (doc)
        {
            JsonElement entries;
            if (doc.RootElement.ValueKind == JsonValueKind.Array)
            {
                entries = doc.RootElement;
            }
            else if (doc.RootElement.TryGetProperty("entity", out var entity)
                && entity.ValueKind == JsonValueKind.Array)
            {
                entries = entity;
            }
            else
            {
                throw new ParticipantNotFoundException(
                    $"Registry response had no entry for {participantCode}");
            }

            if (entries.GetArrayLength() == 0)
            {
                throw new ParticipantNotFoundException(
                    $"Registry response had no entry for {participantCode}");
            }

            var first = entries[0];
            if (!first.TryGetProperty("encryption_cert", out var certElement)
                || certElement.ValueKind != JsonValueKind.String)
            {
                throw new TransportException(
                    $"Registry entry for {participantCode} has no encryption_cert");
            }

            var pem = certElement.GetString()!;
            return ParseCert(participantCode, pem);
        }
    }

    private static ParticipantCert ParseCert(string participantCode, string pem)
    {
        X509Certificate2 cert;
        try
        {
            cert = X509Certificate2.CreateFromPem(pem);
        }
        catch (CryptographicException ex)
        {
            throw new TransportException(
                $"Failed to parse encryption_cert PEM for {participantCode}: {ex.Message}");
        }

        try
        {
            var rsa = cert.GetRSAPublicKey();
            if (rsa is null)
            {
                cert.Dispose();
                throw new TransportException(
                    $"encryption_cert for {participantCode} is not RSA");
            }

            return new ParticipantCert(
                ParticipantCode: participantCode,
                PublicKey: rsa,
                NotAfter: cert.NotAfter.ToUniversalTime());
        }
        finally
        {
            cert.Dispose();
        }
    }

    private void MaybeLogStats()
    {
        DateTimeOffset now;
        bool shouldLog;
        long hits, misses, evictions;

        lock (_statsLock)
        {
            now = _clock();
            shouldLog = (now - _lastStatsLog) >= TimeSpan.FromSeconds(60);
            if (shouldLog)
            {
                _lastStatsLog = now;
            }

            hits = Interlocked.Read(ref _hits);
            misses = Interlocked.Read(ref _misses);
            evictions = Interlocked.Read(ref _evictions);
        }

        if (shouldLog)
        {
            _log.LogInformation(
                "registry cache stats: hits={Hits} misses={Misses} evictions={Evictions}",
                hits,
                misses,
                evictions);
        }
    }

    /// <inheritdoc />
    public void Dispose()
    {
        _cache.Dispose();
        if (_ownsHttpClient)
        {
            _httpClient.Dispose();
        }
    }
}
